trl/data_utils.py

```diff
@@ -466,6 +471,132 @@ def pack_examples(examples: dict[str, list[list]], seq_length: int) -> dict[str,
     return examples
 
 
+def pack_dataset(dataset: DatasetType, seq_length: int, map_kwargs: Optional[dict[str, Any]] = None) -> DatasetType:
+    r"""
+    Pack sequences in a dataset into chunks of size `seq_length`.
+
+    Args:
+        dataset (`Dataset` or `DatasetDict`):
+            Dataset to pack
+        seq_length (`int`):
+            Target sequence length to pack to.
+        map_kwargs (`dict` or `None`, *optional*, defaults to `None`):
+            Additional keyword arguments to pass to the dataset's map method when packing examples.
+
+    Returns:
+        `Dataset` or `DatasetDict`: The dataset with packed sequences. The number of examples may
+        decrease as sequences are combined.
+
+    Example:
+    ```python
+    >>> from datasets import Dataset
+    >>> examples = {
+    ...     "input_ids": [[1, 2], [3, 4], [5, 6], [7]],
+    ...     "attention_mask": [[1, 1], [0, 1], [1, 1], [1]],
+    ... }
+    >>> dataset = Dataset.from_dict(examples)
+    >>> packed_dataset = pack_dataset(dataset, seq_length=4)
+    >>> packed_dataset[:]
+    {'input_ids': [[1, 2, 3, 4], [5, 6, 7]],
+     'attention_mask': [[1, 1, 0, 1], [1, 1, 1]]}
+    ```
+    """
+    if map_kwargs is None:
+        map_kwargs = {}
+    if isinstance(dataset, Dataset):
+        # Fast packing with pyarrow
+        def pack(examples):
+            packed_columns = []
+            for column in examples.columns:
+                if pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type):
+                    if isinstance(column, pa.ChunkedArray):
+                        column = column.combine_chunks()
+                    num_elements = len(column.values)
+                    dtype = column.offsets.type.to_pandas_dtype()  # np.int32 or np.int64
+                    offsets = np.arange(0, num_elements + 1, seq_length, dtype=dtype)
+                    if offsets[-1] != num_elements:
+                        offsets = np.concatenate([offsets, [num_elements]])
+                    column = type(column).from_arrays(offsets, column.values)
```

Comment:
Is there a `pyarrow.compute` function you can use here instead to simplify this ?

The functions from `pyarrow.compute` generally do a copy of the data, but since `.combine_chunks()` also copies the data it might not affect performance that much

Reply:
The compute API is very limited, so I think this is the best we can do :)

The only way to avoid the copy would be by packing the individual `.chunks` (without crossing the boundaries), but this would then make the operation dependent on the underlying chunking and lead to more padding afterward, so I think the copy is justified here.

Reply:
Made the change to handle that, but I don't remember in what scenario we receive sliced arrays as input, so without a test case.

Reply:
lgtm ! I think it can happen when training on a small portion of the train set, like `--dataset_train_split "train[:100]"`